File: fritzStreams/src/main/java/de/acepe/fritzstreams/ui/components/DownloadEntryView.java
package de.acepe.fritzstreams.ui.components;

import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.net.Uri;
import android.support.v4.content.FileProvider;
import android.util.AttributeSet;
import android.view.View;
import android.widget.*;
import de.acepe.fritzstreams.BuildConfig;
import de.acepe.fritzstreams.R;
import de.acepe.fritzstreams.backend.DownloadInfo;
import de.acepe.fritzstreams.backend.DownloadState;

import java.io.File;

import static de.acepe.fritzstreams.backend.DownloadState.DOWNLOADING;
import static de.acepe.fritzstreams.backend.DownloadState.FINISHED;

public class DownloadEntryView extends LinearLayout {

    public interface Action {
        void execute(DownloadInfo t);
    }

    public static final String TAG = "StreamDownloadView";
    private TextView mTitle;
    private TextView mSubTitle;
    private TextView mState;
    private TextView mPlay;
    private Button mCancelButton;
    private ProgressBar mDownloadProgress;
    private DownloadInfo mDownload;

    public DownloadEntryView(Context context) {
        super(context);
        init(context);
    }

    public DownloadEntryView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public DownloadEntryView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context);
    }

    private void init(Context context) {
        Typeface font = Typeface.createFromAsset(context.getAssets(), "fontawesome-webfont.ttf");
        View.inflate(context, R.layout.download_entry_view, this);

        mTitle = findViewById(R.id.tvTitle);
        mSubTitle = findViewById(R.id.tvSubtitle);
        mState = findViewById(R.id.tvState);
        mPlay = findViewById(R.id.tvPlay);
        mPlay.setTypeface(font);
        mCancelButton = findViewById(R.id.btnCancelDownload);
        mCancelButton.setTypeface(font);
        mDownloadProgress = findViewById(R.id.pbDownloadProgress);
        setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mDownload.getState() == DownloadState.FINISHED) {
                    DownloadEntryView.this.openWithMusicApp();
                }
            }
        });
    }

    public void setButtonAction(final Action action) {
        mCancelButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                action.execute(mDownload);
            }
        });
    }

    public void setDownload(DownloadInfo download) {
        mDownload = download;
        mTitle.setText(download.getTitle());
        mSubTitle.setText(download.getSubtitle());

        DownloadState state = download.getState();
        switch (state) {
            case WAITING:
                mCancelButton.setText(R.string.icon_remove);
                mState.setText(R.string.waiting);
                break;
            case DOWNLOADING:
                mDownloadProgress.setProgress(download.getProgressPercent());
                break;
            case CANCELLED:
                mState.setText(R.string.cancelled);
                mCancelButton.setText(R.string.icon_retry);
                break;
            case FAILED:
                mState.setText(R.string.failed);
                mCancelButton.setText(R.string.icon_retry);
                break;
            case FINISHED:
                mState.setText(R.string.finished);
                mCancelButton.setText(R.string.icon_accept);
                break;
        }
        mDownloadProgress.setVisibility(state == DOWNLOADING ? View.VISIBLE : View.INVISIBLE);
        mState.setVisibility(state == DOWNLOADING ? View.INVISIBLE : View.VISIBLE);
        mPlay.setVisibility(state == FINISHED ? View.VISIBLE : View.INVISIBLE);
    }

    public DownloadInfo getDownload() {
        return mDownload;
    }

    private void openWithMusicApp() {
        Uri outFile = FileProvider.getUriForFile(getContext(), BuildConfig.APPLICATION_ID + ".provider", new File(mDownload.getFilename()));

        Intent mediaIntent = new Intent();
        mediaIntent.setAction(Intent.ACTION_VIEW);
        mediaIntent.setDataAndType(outFile, "audio/*");
        mediaIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_GRANT_READ_URI_PERMISSION);

        if (mediaIntent.resolveActivity(getContext().getPackageManager()) != null) {
            getContext().startActivity(mediaIntent);
        } else {
            Toast.makeText(getContext(), R.string.app_not_available, Toast.LENGTH_LONG).show();
        }
    }
}

File: fritzStreams/src/main/java/de/acepe/fritzstreams/ui/components/DownloadEntryView.kt
package de.acepe.fritzstreams.ui.components

import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.support.v4.content.FileProvider
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import de.acepe.fritzstreams.BuildConfig
import de.acepe.fritzstreams.R
import de.acepe.fritzstreams.backend.DownloadInfo
import de.acepe.fritzstreams.backend.DownloadState.*
import kotlinx.android.synthetic.main.download_entry_view.view.*
import java.io.File

private const val FONTAWESOME = "fontawesome-webfont.ttf"

class DownloadEntryView(context: Context) : LinearLayout(context) {

    private lateinit var mDownload: DownloadInfo

    init {
        val font = Typeface.createFromAsset(context.assets, FONTAWESOME)
        LayoutInflater.from(context).inflate(R.layout.download_entry_view, this)

        playIcon.typeface = font
        cancelDownload.typeface = font
    }

    fun setButtonAction(action: (DownloadInfo) -> Unit) {
        cancelDownload.setOnClickListener { action(mDownload) }
    }

    fun setDownload(download: DownloadInfo) {
        mDownload = download
        title.text = download.title
        subtitle.text = download.subtitle

        when (download.state) {
            WAITING -> updateView(R.string.waiting, R.string.icon_remove)
            DOWNLOADING -> downloadProgress.progress = download.progressPercent
            CANCELLED -> updateView(R.string.cancelled, R.string.icon_retry)
            FAILED -> updateView(R.string.failed, R.string.icon_retry)
            FINISHED -> updateView(R.string.finished, R.string.icon_accept)
        }
        downloadProgress.visibility = if (download.state == DOWNLOADING) View.VISIBLE else View.INVISIBLE
        stateDisplay.visibility = if (download.state == DOWNLOADING) View.INVISIBLE else View.VISIBLE
        playIcon.visibility = if (download.state == FINISHED) View.VISIBLE else View.INVISIBLE

        setOnClickListener { if (download.state == FINISHED) openWithMusicApp() }
    }

    private fun updateView(stateText: Int, cancelText: Int) {
        stateDisplay.setText(stateText)
        cancelDownload.setText(cancelText)
    }

    private fun openWithMusicApp() {
        val outFile = FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID + ".provider", File(mDownload.filename))

        val mediaIntent = Intent()
        with(mediaIntent) {
            action = Intent.ACTION_VIEW
            setDataAndType(outFile, "audio/*")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }

        if (mediaIntent.resolveActivity(context.packageManager) != null) {
            context.startActivity(mediaIntent)
        } else {
            Toast.makeText(context, R.string.app_not_available, Toast.LENGTH_LONG).show()
        }
    }

}